# Introduction

In this NoteBook, we will clean the dataset of "charging station"

## Set up

In [1]:
import pandas as pd
import numpy as np
import csv
from datetime import datetime

## Read the file

In [8]:
#read the file
path = '../raw/charging.csv'
df = pd.read_csv(path)

D:\Anaconda\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (9) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [4]:
df.head()

Fuel Type Code      Fuel Type                                  Station Name  \
0            E85  Ethanol (E85)                                Kwik Trip #937   
1           ELEC       Electric                      Hollywood Plaza Port 1 2   
2           ELEC       Electric                 Dwntwn Stations Law Ll2 North   
3           ELEC       Electric  The Shops At La Cantera - Cheesecake Factory   
4           ELEC       Electric                   Endeavor Reg 901 E 6Th St 3   

                                  Street Address Intersection Directions  \
0                              2550 Highway 12 E                     NaN   
1                          8050 International Dr                     NaN   
2  Law School Underground Garage 111 N First Str                     NaN   
3                          15900 La Cantera Pkwy                     NaN   
4                                   901 E 6th St                     NaN   

          City State Code RegionId      State    ZIP  ...  \
0      Willmar         MN    US_MN  Minnesota  56201  ...   
1      Orlando         FL    US_FL    Florida  32819  ...   
2      Phoenix         AZ    US_AZ    Arizona  85004  ...   
3  San Antonio         TX    US_TX      Texas  78256  ...   
4       Austin         TX    US_TX      Texas  78702  ...   

   E85 Other Ethanol Blends EV Pricing LPG Nozzle Types Hydrogen Pressures  \
0                   ["E15"]        NaN              NaN                NaN   
1                       NaN        NaN              NaN                NaN   
2                       NaN        NaN              NaN                NaN   
3                       NaN        NaN              NaN                NaN   
4                       NaN        NaN              NaN                NaN   

  Hydrogen Standards CNG Fill Type Code CNG PSI  CNG Vehicle Class  \
0                NaN                NaN     NaN                NaN   
1                NaN                NaN     NaN                NaN   
2                NaN                NaN     NaN                NaN   
3                NaN                NaN     NaN                NaN   
4                NaN                NaN     NaN                NaN   

  LNG Vehicle Class EV On-Site Renewable Source  
0               NaN                         NaN  
1               NaN                         NaN  
2               NaN                         NaN  
3               NaN                         NaN  
4               NaN                         NaN  

[5 rows x 65 columns]

#### There are too many data(each row corresponding to the information of a station), So we need to clean it. 
#### The columns we need is state of station, and their open date

In [5]:
list(df)

['Fuel Type Code',
 'Fuel Type',
 'Station Name',
 'Street Address',
 'Intersection Directions',
 'City',
 'State Code',
 'RegionId',
 'State',
 'ZIP',
 'Plus4',
 'Station Phone',
 'Status Code',
 'Status',
 'Status note',
 'Expected Date',
 'Groups With Access Code',
 'Groups with access code note',
 'Access Days Time',
 'Cards Accepted',
 'BD Blends',
 'NG Fill Type Code',
 'NG PSI',
 'EV Level1 EVSE Num',
 'EV Level2 EVSE Num',
 'EV DC Fast Count',
 'EV Other Info',
 'EV Network',
 'EV Network Web',
 'Geocode Status',
 'Latitude',
 'Longitude',
 'Date Last Confirmed',
 'ID',
 'Updated At',
 'Owner Type Code',
 'Federal Agency ID',
 'Federal Agency Name',
 'Open Date',
 'Hydrogen Status Link',
 'NG Vehicle Class',
 'LPG Primary',
 'E85 Blender Pump',
 'EV Connector Types',
 'Country',
 'Groups With Access Code (French)',
 'Hydrogen Is Retail',
 'Access Code',
 'Access Detail Code',
 'Federal Agency Code',
 'Facility Type',
 'CNG Dispenser Num',
 'CNG On-Site Renewable Source',
 'CNG 

In [6]:
## drop the columns we don't need
df = df.drop(['Fuel Type Code',
 'Fuel Type',
 'Station Name',
 'Street Address',
 'Intersection Directions',
 'City',
 'State Code',
 'RegionId',
 'ZIP',
 'Plus4',
 'Station Phone',
 'Status Code',
 'Status',
 'Status note',
 'Expected Date',
 'Groups With Access Code',
 'Groups with access code note',
 'Access Days Time',
 'Cards Accepted',
 'BD Blends',
 'NG Fill Type Code',
 'NG PSI',
 'EV Level1 EVSE Num',
 'EV Level2 EVSE Num',
 'EV DC Fast Count',
 'EV Other Info',
 'EV Network',
 'EV Network Web',
 'Geocode Status',
 'Latitude',
 'Longitude',
 'Date Last Confirmed',
 'ID',
 'Updated At',
 'Owner Type Code',
 'Federal Agency ID',
 'Federal Agency Name',
 'Hydrogen Status Link',
 'NG Vehicle Class',
 'LPG Primary',
 'E85 Blender Pump',
 'EV Connector Types',
 'Country',
 'Groups With Access Code (French)',
 'Hydrogen Is Retail',
 'Access Code',
 'Access Detail Code',
 'Federal Agency Code',
 'Facility Type',
 'CNG Dispenser Num',
 'CNG On-Site Renewable Source',
 'CNG Total Compression Capacity',
 'CNG Storage Capacity',
 'E85 Other Ethanol Blends',
 'EV Pricing',
 'LPG Nozzle Types',
 'Hydrogen Pressures',
 'Hydrogen Standards',
 'CNG Fill Type Code',
 'CNG PSI',
 'CNG Vehicle Class',
 'LNG Vehicle Class',
 'EV On-Site Renewable Source'],axis=1)

In [7]:
df.head()

State   Open Date
0  Minnesota  09/15/2018
1    Florida  04/03/2019
2    Arizona  10/22/2019
3      Texas  11/23/2020
4      Texas  04/02/2021

In [8]:
# we just need the data in year, not exactly date. So clean it
df['Open Date'] = pd.to_datetime(df['Open Date'])

df['Open Date'] = df['Open Date'].dt.year

In [9]:
df

State  Open Date
0       Minnesota     2018.0
1         Florida     2019.0
2         Arizona     2019.0
3           Texas     2020.0
4           Texas     2021.0
...           ...        ...
60466   Minnesota     2020.0
60467  California     2020.0
60468        Utah     2021.0
60469    Nebraska     2021.0
60470       Texas     2021.0

[60471 rows x 2 columns]

In [10]:
# there are some data of 'Open Date' is not avaliable, so we need to clean it
# fill the 'NA' as value 0
df = df.fillna(0)
#convert the data type to int
df['Open Date'] = df['Open Date'].astype(int)
#clean the unavailable data
list = []
for index, row in df.iterrows():
    if df.iloc[index,1] == 0:
        list.append(index)
df = df.drop(list,axis=0)
df

State  Open Date
0       Minnesota       2018
1         Florida       2019
2         Arizona       2019
3           Texas       2020
4           Texas       2021
...           ...        ...
60466   Minnesota       2020
60467  California       2020
60468        Utah       2021
60469    Nebraska       2021
60470       Texas       2021

[60124 rows x 2 columns]

## Now we have data of all 60470 stations' information, we need to clean them as:
## "The number of stations for each state by year"

In [10]:
#create an array to store the year
a = np.sort(df['Open Date'].unique())
a

array([1974, 1976, 1978, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991,
       1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002,
       2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013,
       2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022])

In [11]:
#creat a list to store the columns name
col = []
col.append('state')
for i in a:
    col.append(i)

In [12]:
# creat a new DataFrame that ready to get value for "The number of stations for each state by year"
df1 = pd.DataFrame(columns = col)
df1['state'] = df['State'].unique()
df1 = df1.fillna(0)
df1 = df1.set_index('state')
df1.head()

1974  1976  1978  1984  1985  1986  1987  1988  1989  1990  ...  \
state                                                                  ...   
Minnesota     0     0     0     0     0     0     0     0     0     0  ...   
Florida       0     0     0     0     0     0     0     0     0     0  ...   
Arizona       0     0     0     0     0     0     0     0     0     0  ...   
Texas         0     0     0     0     0     0     0     0     0     0  ...   
Illinois      0     0     0     0     0     0     0     0     0     0  ...   

           2013  2014  2015  2016  2017  2018  2019  2020  2021  2022  
state                                                                  
Minnesota     0     0     0     0     0     0     0     0     0     0  
Florida       0     0     0     0     0     0     0     0     0     0  
Arizona       0     0     0     0     0     0     0     0     0     0  
Texas         0     0     0     0     0     0     0     0     0     0  
Illinois      0     0     0     0     0     0     0     0     0     0  

[5 rows x 42 columns]

In [13]:
print(df['State'].unique())

['Minnesota' 'Florida' 'Arizona' 'Texas' 'Illinois' 'California'
 'North Carolina' 'Ohio' 'Georgia' 'Massachusetts' 'Kentucky' 'Michigan'
 'Iowa' 'Washington' 'New York' 'West Virginia' 'Colorado' 'Idaho'
 'Nevada' 'Kansas' 'Wisconsin' 'Indiana' 'Pennsylvania' 'Maryland'
 'South Carolina' 'Oklahoma' 'South Dakota' 'Tennessee' 'Alabama' 'Utah'
 'New Jersey' 'Oregon' 'Missouri' 'New Mexico' 'Virginia' 'Mississippi'
 'Maine' 'Vermont' 'Delaware' 'Louisiana' 'Rhode Island' 'Montana'
 'Connecticut' 'Hawaii' 'Nebraska' 'Arkansas' 'Ontario'
 'District of Columbia' 'New Hampshire' 'Alaska' 'Wyoming' 'North Dakota'
 'Puerto Rico']


In [14]:
# iterrate the dataframe, using condition to count the number of stations then put it into new DataFrame
for state in df['State'].unique():
    for year in a:
        count=0
        for index, row in df.iterrows():
            if df['State'][index] == state and df['Open Date'][index] <= year:
                count += 1
        df1.at[state,year] = count

In [15]:
df1

1974  1976  1978  1984  1985  1986  1987  1988  1989  \
state                                                                        
Minnesota                0     0     0     0     0     0     0     0     0   
Florida                  0     0     0     0     0     0     0     0     0   
Arizona                  0     0     0     1     1     1     1     1     1   
Texas                    0     0     0     0     0     0     0     0     0   
Illinois                 0     0     0     0     0     0     0     0     0   
California               0     0     0     0     0     0     0     0     0   
North Carolina           0     0     0     0     0     0     0     0     0   
Ohio                     0     1     1     1     1     2     2     2     2   
Georgia                  0     0     0     0     0     0     1     1     1   
Massachusetts            0     0     0     0     0     0     0     0     0   
Kentucky                 0     0     0     0     0     0     0     0     0   
Michigan                 0     0     0     0     0     0     0     0     0   
Iowa                     0     0     0     0     0     0     0     1     1   
Washington               0     0     0     0     1     1     1     1     1   
New York                 0     0     0     0     0     0     0     1     1   
West Virginia            0     0     0     0     0     0     0     0     0   
Colorado                 0     0     0     0     0     0     0     0     0   
Idaho                    0     0     0     0     0     1     2     3     4   
Nevada                   0     0     0     0     0     0     0     0     0   
Kansas                   0     0     0     0     0     0     0     0     0   
Wisconsin                0     0     0     0     0     0     0     0     0   
Indiana                  0     0     0     0     0     0     0     0     0   
Pennsylvania             0     0     0     0     0     0     0     0     0   
Maryland                 0     0     1     1     1     1     1     1     1   
South Carolina           0     0     0     0     0     0     0     0     1   
Oklahoma                 0     0     0     0     0     0     0     0     0   
South Dakota             0     0     0     0     0     0     0     0     0   
Tennessee                0     0     0     0     0     0     0     0     0   
Alabama                  0     0     0     0     0     0     0     0     0   
Utah                     0     0     0     0     0     0     0     0     0   
New Jersey               0     0     0     0     0     0     0     0     0   
Oregon                   0     0     0     0     0     0     0     0     0   
Missouri                 0     0     0     0     0     0     0     0     0   
New Mexico               0     0     0     0     0     0     0     0     0   
Virginia                 0     0     0     0     0     0     0     0     0   
Mississippi              0     0     0     0     0     0     0     0     0   
Maine                    0     0     0     0     0     0     0     0     0   
Vermont                  0     0     0     0     0     0     0     0     0   
Delaware                 0     0     0     0     0     0     0     0     0   
Louisiana                0     0     0     0     0     0     0     0     0   
Rhode Island             0     0     0     0     0     0     0     0     0   
Montana                  0     0     0     0     0     0     0     0     0   
Connecticut              0     0     0     0     0     0     0     0     0   
Hawaii                   0     0     0     0     0     0     0     0     0   
Nebraska                 0     0     0     0     1     1     2     2     2   
Arkansas                 0     0     0     0     1     1     1     1     1   
Ontario                  0     0     0     0     0     0     0     0     0   
District of Columbia     0     0     0     0     0     0     0     0     0   
New Hampshire            0     0     0     0     0     0     0     0     0   
Alaska                   1     1     1     1

In [16]:
# Now we have cleanned data, save it into processed file
path = '../processed/charging_station.csv'
df1.to_csv(path)